// openshift-express/lib/index.js
/**
 * openshift-express
 * https://github.com/magdev/openshift-express
 *
 * Copyright (c) 2014 Marco Graetsch
 * Licensed under the MIT license.
 */

'use strict';

var extend = require('xtend'),

    /** Default Configuration */
    defaults = {
        server: {
            hostname: '0.0.0.0',
            port: 3000,
            ipaddr: '0.0.0.0'
        },
        databases: []
    },
    
    /** Default Database Configuration */
    databaseDefaults = {
        mysql: {
            host: 'localhost',
            port: 3306,
            username: 'root',
            password: null,
            database: 'database'
        },
        mongodb: {
            host: 'localhost',
            port: 27017,
            username: null,
            password: null,
            database: 'database'
        }
    },
    
    
    /**
     * Build the DSN for a database
     * 
     * @param {Object} data
     * @returns {String}
     */
    buildDSN = function(data) {
        var type = data.type.toUpperCase(),
            defaults = databaseDefaults[data.type],
            
            config = extend(data, {
                host: (process.env['OPENSHIFT_' + type + '_DB_HOST'] || defaults.host),
                port: (process.env['OPENSHIFT_' + type + '_DB_PORT'] || defaults.port),
                username: (process.env['OPENSHIFT_' + type + '_DB_USERNAME'] || defaults.username),
                password: (process.env['OPENSHIFT_' + type + '_DB_PASSWORD'] || defaults.password),
                database: (process.env.OPENSHIFT_APP_NAME || defaults.database)
            });

        if (config.username) {
            return data.type + '://' + config.username + (config.password ? ':' + config.password : '') + '@' + config.host + ':' + config.port + '/' + config.database;
        }
        return data.type + '://' + config.host + ':' + config.port + '/' + config.database;
    };
    

/**
 * Main module
 * 
 * @param {Application} app
 * @param {String} dbType
 */
module.exports = exports = function(app, config) {
    var cfg = extend(defaults, config);
    
    app.set('hostname', (process.env.OPENSHIFT_APP_DNS || process.env.HOST || cfg.server.host || '0.0.0.0'));
    app.set('port', (process.env.OPENSHIFT_NODEJS_PORT || cfg.server.port || 3000));
    app.set('ipaddr', (process.env.OPENSHIFT_NODEJS_IP  || cfg.server.ipaddr || '0.0.0.0'));
    
    app.set('openshift', (process.env.OPENSHIFT_APP_NAME ? true : false));
    if (app.get('openshift')) { 
        app.set('openshift appname', process.env.OPENSHIFT_APP_NAME);
    }
    
    if (cfg.databases && cfg.databases.length) {
        cfg.databases.forEach(function(data) {
            app.set(data.type + ' dsn', buildDSN(data));
        });
    }
};
